toqito/matrices/operator_sinkhorn.py

```diff
@@ -0,0 +1,198 @@
+
+    # precompute trace for all iterations
+    tr_rho_p = np.power(tr_rho, 1.0 / (2.0 * num_sys)) if tr_rho != 0 else 1.0
+
+    # Prepare the iteration.
+    Prho = [np.eye(d, dtype=np.complex128) / d for d in dim]  # Force complex128 type for Prho
+    F = [np.eye(d, dtype=np.complex128) * tr_rho_p for d in dim]  # Force complex128 for F
+    ldim = [np.prod(dim[:j]) for j in range(num_sys)]
+    rdim = [np.prod(dim[j+1:]) for j in range(num_sys)]
+
+    # Perform the operator Sinkhorn iteration.
+    it_err = 1.0
+    max_cond = 0
+    max_iterations = 10000
+    iterations = 0
+    rho2 = rho.copy().astype(np.complex128)
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        while it_err > tol:
+
+            if iterations > max_iterations:
+                raise RuntimeError(f"operator_sinkhorn did not converge within {max_iterations} iterations.")
+
+            it_err = 0.0
+            max_cond = 0.0
+
+            error_flag_in_iteration = False
+
+            try:
+                for j in range(num_sys):
+                    # Compute the reduced density matrix on the j-th system.
+                    Prho_tmp = partial_trace(rho2, list(set(range(num_sys)) - {j}), dim)
+                    Prho_tmp = (Prho_tmp + Prho_tmp.conj().T) / 2.0  # for numerical stability
+                    it_err += np.linalg.norm(Prho[j] - Prho_tmp)
+                    Prho[j] = Prho_tmp.astype(np.complex128)  # Force complex128 for Prho_tmp
+
+                    # Apply the filter with explicit complex128 conversion
+                    T_inv = np.linalg.inv(Prho[j]).astype(np.complex128)
+
+                    # check for NaNs and infinities after inversion
+                    if np.any(np.isnan(T_inv)) or np.any(np.isinf(T_inv)):
+                        error_flag_in_iteration = True
+
+                    T = scipy.linalg.sqrtm(T_inv) / np.sqrt(dim[j])
+                    T = T.astype(np.complex128)
+
+                    # enforce hermiticity for numerical stability
+                    T = (T + T.conj().T) / 2.0
+
+                    # Construct the Kronecker product
+                    eye_ldim = np.eye(int(ldim[j]), dtype=np.complex128)
+                    eye_rdim = np.eye(int(rdim[j]), dtype=np.complex128)
+                    Tk = np.kron(eye_ldim, np.kron(T, eye_rdim))
+
+                    rho2 = Tk @ rho2 @ Tk.conj().T
+                    F[j] = T @ F[j]
+
+                    max_cond = max(max_cond, np.linalg.cond(F[j]))
+
+            except Exception:
+                error_flag_in_iteration = True
+
+            # Check the condition number to ensure invertibility.
+            if (error_flag_in_iteration) or (max_cond >= 1 / tol) or (np.isinf(max_cond)):
+                raise ValueError("The operator Sinkhorn iteration does not converge for RHO. "
+                             "This is often the case if RHO is not of full rank.")
```

Comment:
Seems like there is some odd indentation going on with the multi-line strings here.